configfs.c

```diff
@@ -186,6 +186,17 @@ char *tcmu_get_cfgfs_str(const char *path)
 	if (ret == 0)
 		return NULL;
 
+	/*
+	 * Some files like members will with a null char at the end of each
+	 * member, just replace it with '\n', except the last one.
+	 */
+	if (ret != strlen(buf)) {
+		do {
+			n = strlen(buf);
+			buf[n] = '\n';
+		} while (n < ret);
+	}
+
 	/*
```

Comment:
Are you able to hit this? I thought the kernel has a bug where it is only allowing one member. Are you able to configure multiple members and if so how and what kernel?

Reply:
If you can add multiple members could you send me the

targetcli ls

output and the output of the members file?

Reply:
Hi Mike,
Yes we can hit this for glfs block currently. please see the targetcli ls output:

$ targetcli ls
o- / ......................................................................................................................... [...]
  o- backstores .............................................................................................................. [...]
  | o- block .................................................................................................. [Storage Objects: 0]
  | o- fileio ................................................................................................. [Storage Objects: 0]
  | o- pscsi .................................................................................................. [Storage Objects: 0]
  | o- ramdisk ................................................................................................ [Storage Objects: 0]
  | o- user:glfs .............................................................................................. [Storage Objects: 1]
  | | o- block300 ....................... [dht0@192.168.195.134/block-store/0904c613-d6d6-4089-9f74-458344840863 (3.0GiB) activated]
  | |   o- alua ................................................................................................... [ALUA Groups: 1]
  | |     o- default_tg_pt_gp ....................................................................... [ALUA state: Active/optimized]
  | o- user:qcow .............................................................................................. [Storage Objects: 0]
  | o- user:rbd ............................................................................................... [Storage Objects: 0]
  o- iscsi ............................................................................................................ [Targets: 1]
  | o- iqn.2016-12.org.gluster-block:0904c613-d6d6-4089-9f74-458344840863 ................................................ [TPGs: 3]
  |   o- tpg1 ........................................................................................................... [disabled]
  |   | o- acls .......................................................................................................... [ACLs: 0]
  |   | o- luns .......................................................................................................... [LUNs: 1]
  |   | | o- lun0 ............................................................................... [user/block300 (default_tg_pt_gp)]
  |   | o- portals .................................................................................................... [Portals: 1]
  |   |   o- 192.168.195.137:3260 ............................................................................................. [OK]
  |   o- tpg2 .................................................................................................. [gen-acls, no-auth]
  |   | o- acls .......................................................................................................... [ACLs: 0]
  |   | o- luns .......................................................................................................... [LUNs: 1]
  |   | | o- lun0 ............................................................................... [user/block300 (default_tg_pt_gp)]
  |   | o- portals .................................................................................................... [Portals: 1]
  |   |   o- 192.168.195.134:3260 ............................................................................................. [OK]
  |   o- tpg3 ........................................................................................................... [disabled]
  |     o- acls .......................................................................................................... [ACLs: 0]
  |     o- luns .......................................................................................................... [LUNs: 1]
  |     | o- lun0 ............................................................................... [user/block300 (default_tg_pt_gp)]
  |     o- portals .................................................................................................... [Portals: 1]
  |       o- 192.168.195.135:3260 ............................................................................................. [OK]
  o- loopback ......................................................................................................... [Targets: 0]
  o- vhost ............................................................................................................ [Targets: 0]
  o- xen-pvscsi ....................................................................................................... [Targets: 0]


$  cat /sys/kernel/config/target/core/user_0/block300/alua/default_tg_pt_gp/members 
iSCSI/iqn.2016-12.org.gluster-block:0904c613-d6d6-4089-9f74-458344840863/tpgt_3/lun_0
iSCSI/iqn.2016-12.org.gluster-block:0904c613-d6d6-4089-9f74-458344840863/tpgt_2/lun_0
iSCSI/iqn.2016-12.org.gluster-block:0904c613-d6d6-4089-9f74-458344840863/tpgt_1/lun_0


$ uname -a
Linux gblock2 3.10.0-693.5.2.3.el7.centos.x86_64 #1 SMP Tue Jan 16 16:38:59 CST 2018 x86_64 x86_64 x86_64 GNU/Linux

This kernel is 693 with your kernel patches you gave to me.

Reply:
Huh, for some reason I could not get that to work before. Maybe due to the null/newline parsing issue.

There might be some bugs because I have not tested multiple member setups.

At the version least some comments are bogus now. Like the comment in tcmu_get_enabled_port should say we support the setup you are doing where you have one group with multiple ports and the local one enabled.

Also, I think  in tcmu_add_dev_to_recovery_list we need to fix up the members comment and the loops in there should be fixed to account for there being multiple tgt_ports like this:

[runner-break-fron-loop.txt](https://github.com/open-iscsi/tcmu-runner/files/1800881/runner-break-fron-loop.txt)